notebooks/tutorials/beginner_workflow/7_Retrieving_remote_data.ipynb

```diff
@@ -0,0 +1,399 @@
+    "\n",
+    "You can see more information about the instruments by going to the link in the `instrument_data` section of the metadata"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "a302229c",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "metadata = wao_data.metadata\n",
+    "instrument_data = metadata[\"instrument_data\"]\n",
+    "citation_string = metadata[\"citation_string\"]"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "e2c38cba",
+   "metadata": {},
+   "source": [
+    "Here we get the instrument name and a link to the instrument data on the ICOS Carbon Portal."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "eb6c21f8",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "instrument_data"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "468f10c9",
+   "metadata": {},
+   "source": [
+    "And we can easily get the citation string for the data"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "ae981e6b",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "citation_string"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "29b6217c",
+   "metadata": {},
+   "source": [
+    "### Viewing the data\n",
+    "\n",
+    "As with any `ObsData` object we can quickly plot it to have a look.\n",
+    "\n",
+    "> **_NOTE:_**  the plot created below may not show up on the online documentation version of this notebook."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "c75605a5",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "wao_data.plot_timeseries()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "730bb2fd",
+   "metadata": {},
+   "source": [
+    "### Data levels\n",
+    "\n",
+    "Data available on the ICOS Carbon Portal is made available under three different levels ([see docs](https://icos-carbon-portal.github.io/pylib/modules/#stationdatalevelnone)).\n",
+    "\n",
+    "```\n",
+    "- Data level 1: Near Real Time Data (NRT) or Internal Work data (IW).\n",
+    "- Data level 2: The final quality checked ICOS RI data set, published by the CFs,\n",
+    "                to be distributed through the Carbon Portal.\n",
+    "                This level is the ICOS-data product and free available for users.\n",
+    "- Data level 3: All kinds of elaborated products by scientific communities\n",
+    "                that rely on ICOS data products are called Level 3 data.\n",
+    "```\n",
+    "\n",
+    "By default level 2 data is retrieved but this can be changed by passing `data_level` to `retrieve_icos`. Below we'll retrieve some more recent data from **WAO**."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "59f48a55",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "wao_data_level1 = retrieve_icos(site=\"WAO\", species=\"CH4\", data_level=1)"
```

Comment:
So at the moment, if I try running this notebook I get the following error:
```
KeyError                                  Traceback (most recent call last)
~/work/environments/openghg_env/lib/python3.8/site-packages/xarray/core/dataset.py in _construct_dataarray(self, name)
   1392         try:
-> 1393             variable = self._variables[name]
   1394         except KeyError:

KeyError: 'userflag'

During handling of the above exception, another exception occurred:

KeyError                                  Traceback (most recent call last)
~/work/environments/openghg_env/lib/python3.8/site-packages/xarray/core/concat.py in _dataset_concat(datasets, dim, data_vars, coords, compat, positions, fill_value, join, combine_attrs)
    513             try:
--> 514                 vars = ensure_common_dims([ds[k].variable for ds in datasets])
    515             except KeyError:

~/work/environments/openghg_env/lib/python3.8/site-packages/xarray/core/concat.py in <listcomp>(.0)
    513             try:
--> 514                 vars = ensure_common_dims([ds[k].variable for ds in datasets])
    515             except KeyError:

~/work/environments/openghg_env/lib/python3.8/site-packages/xarray/core/dataset.py in __getitem__(self, key)
   1496         if hashable(key):
-> 1497             return self._construct_dataarray(key)
   1498         else:

~/work/environments/openghg_env/lib/python3.8/site-packages/xarray/core/dataset.py in _construct_dataarray(self, name)
   1394         except KeyError:
-> 1395             _, name, variable = _get_virtual_variable(
   1396                 self._variables, name, self._level_coords, self.dims

~/work/environments/openghg_env/lib/python3.8/site-packages/xarray/core/dataset.py in _get_virtual_variable(variables, key, level_vars, dim_sizes)
    167     else:
--> 168         ref_var = variables[ref_name]
    169 

KeyError: 'userflag'

During handling of the above exception, another exception occurred:

ValueError                                Traceback (most recent call last)
/tmp/ipykernel_142148/3349413668.py in <module>
----> 1 wao_data_level1 = retrieve_icos(site="WAO", species="CH4", data_level=1)

~/openghg/openghg/client/_retrieve.py in retrieve_icos(site, species, sampling_height, start_date, end_date, data_level, force_retrieval)
     86         raise NotImplementedError
     87     else:
---> 88         return icos_retrieve(
     89             site=site,
     90             species=species,

~/openghg/openghg/retrieve/icos/_retrieve.py in retrieve(site, species, sampling_height, start_date, end_date, force_retrieval, data_level)
     58 
     59     if results and not force_retrieval:
---> 60         obs_data: Union[ObsData, List[ObsData]] = results.retrieve_all()
     61     else:
     62         # We'll also need to check we have current data

~/openghg/openghg/dataobjects/_searchresults.py in retrieve_all(self)
    192                 for species, inlet_data in species_data.items():
    193                     for inlet in inlet_data:
--> 194                         obsdata = self._create_obsdata(site=site, species=species, inlet=inlet)
    195                         results.append(obsdata)
    196 

~/openghg/openghg/dataobjects/_searchresults.py in _create_obsdata(self, site, species, inlet)
    308             if not self.ranked_data:
    309                 keys = data_keys["unranked"]
--> 310                 final_dataset = recombine_datasets(keys, sort=True)
    311             else:
    312                 dataset_slices = []

~/openghg/openghg/store/_recombination.py in recombine_datasets(keys, sort, attrs_to_check, elevate_inlet)
     85 
     86     # Concatenate datasets along time dimension
---> 87     combined = xr_concat(data, dim="time")
     88 
     89     # Replace/remove incorrect attributes

~/work/environments/openghg_env/lib/python3.8/site-packages/xarray/core/concat.py in concat(objs, dim, data_vars, coords, compat, positions, fill_value, join, combine_attrs)
    236             f"objects, got {type(first_obj)}"
    237         )
--> 238     return f(
    239         objs, dim, data_vars, coords, compat, positions, fill_value, join, combine_attrs
    240     )

~/work/environments/openghg_env/lib/python3.8/site-packages/xarray/core/concat.py in _dataset_concat(datasets, dim, data_vars, coords, compat, positions, fill_value, join, combine_attrs)
    514                 vars = ensure_common_dims([ds[k].variable for ds in datasets])
    515             except KeyError:
--> 516                 raise ValueError(f"{k!r} is not present in all datasets.")
    517             combined = concat_vars(vars, dim, positions, combine_attrs=combine_attrs)
    518             assert isinstance(combined, Variable)

ValueError: 'userflag' is not present in all datasets.
```

Reply:
I think this is similiar to the issue before when creating a new datasource where xarray doesn't like trying to concatentate two datasets that don't have the same data variables. We fixed it before by adding the data variable to the dataset where it's missing but just populating with NaNs e.g. see type of fix in `_datasource.py` file on PR #328 lines 160-181

Reply:
OK I've commented out the lines changing the flag column type for now, we can revisit that, it's not crucial.

Reply:
> So at the moment, if I try running this notebook I get the following error from this cell ("wao_data_level1 = retrieve_icos(site="WAO", species="CH4", data_level=1)"):
> 
> ```
> KeyError                                  Traceback (most recent call last)
> ~/work/environments/openghg_env/lib/python3.8/site-packages/xarray/core/dataset.py in _construct_dataarray(self, name)
>    1392         try:
> -> 1393             variable = self._variables[name]
>    1394         except KeyError:
> 
> KeyError: 'userflag'
> 
> During handling of the above exception, another exception occurred:
> 
> KeyError                                  Traceback (most recent call last)
> ~/work/environments/openghg_env/lib/python3.8/site-packages/xarray/core/concat.py in _dataset_concat(datasets, dim, data_vars, coords, compat, positions, fill_value, join, combine_attrs)
>     513             try:
> --> 514                 vars = ensure_common_dims([ds[k].variable for ds in datasets])
>     515             except KeyError:
> 
> ~/work/environments/openghg_env/lib/python3.8/site-packages/xarray/core/concat.py in <listcomp>(.0)
>     513             try:
> --> 514                 vars = ensure_common_dims([ds[k].variable for ds in datasets])
>     515             except KeyError:
> 
> ~/work/environments/openghg_env/lib/python3.8/site-packages/xarray/core/dataset.py in __getitem__(self, key)
>    1496         if hashable(key):
> -> 1497             return self._construct_dataarray(key)
>    1498         else:
> 
> ~/work/environments/openghg_env/lib/python3.8/site-packages/xarray/core/dataset.py in _construct_dataarray(self, name)
>    1394         except KeyError:
> -> 1395             _, name, variable = _get_virtual_variable(
>    1396                 self._variables, name, self._level_coords, self.dims
> 
> ~/work/environments/openghg_env/lib/python3.8/site-packages/xarray/core/dataset.py in _get_virtual_variable(variables, key, level_vars, dim_sizes)
>     167     else:
> --> 168         ref_var = variables[ref_name]
>     169 
> 
> KeyError: 'userflag'
> 
> During handling of the above exception, another exception occurred:
> 
> ValueError                                Traceback (most recent call last)
> /tmp/ipykernel_142148/3349413668.py in <module>
> ----> 1 wao_data_level1 = retrieve_icos(site="WAO", species="CH4", data_level=1)
> 
> ~/openghg/openghg/client/_retrieve.py in retrieve_icos(site, species, sampling_height, start_date, end_date, data_level, force_retrieval)
>      86         raise NotImplementedError
>      87     else:
> ---> 88         return icos_retrieve(
>      89             site=site,
>      90             species=species,
> 
> ~/openghg/openghg/retrieve/icos/_retrieve.py in retrieve(site, species, sampling_height, start_date, end_date, force_retrieval, data_level)
>      58 
>      59     if results and not force_retrieval:
> ---> 60         obs_data: Union[ObsData, List[ObsData]] = results.retrieve_all()
>      61     else:
>      62         # We'll also need to check we have current data
> 
> ~/openghg/openghg/dataobjects/_searchresults.py in retrieve_all(self)
>     192                 for species, inlet_data in species_data.items():
>     193                     for inlet in inlet_data:
> --> 194                         obsdata = self._create_obsdata(site=site, species=species, inlet=inlet)
>     195                         results.append(obsdata)
>     196 
> 
> ~/openghg/openghg/dataobjects/_searchresults.py in _create_obsdata(self, site, species, inlet)
>     308             if not self.ranked_data:
>     309                 keys = data_keys["unranked"]
> --> 310                 final_dataset = recombine_datasets(keys, sort=True)
>     311             else:
>     312                 dataset_slices = []
> 
> ~/openghg/openghg/store/_recombination.py in recombine_datasets(keys, sort, attrs_to_check, elevate_inlet)
>      85 
>      86     # Concatenate datasets along time dimension
> ---> 87     combined = xr_concat(data, dim="time")
>      88 
>      89     # Replace/remove incorrect attributes
> 
> ~/work/environments/openghg_env/lib/python3.8/site-packages/xarray/core/concat.py in concat(objs, dim, data_vars, coords, compat, positions, fill_value, join, combine_attrs)
>     236             f"objects, got {type(first_obj)}"
>     237         )
> --> 238     return f(
>     239         objs, dim, data_vars, coords, compat, positions, fill_value, join, combine_attrs
>     240     )
> 
> ~/work/environments/openghg_env/lib/python3.8/site-packages/xarray/core/concat.py in _dataset_concat(datasets, dim, data_vars, coords, compat, positions, fill_value, join, combine_attrs)
>     514                 vars = ensure_common_dims([ds[k].variable for ds in datasets])
>     515             except KeyError:
> --> 516                 raise ValueError(f"{k!r} is not present in all datasets.")
>     517             combined = concat_vars(vars, dim, positions, combine_attrs=combine_attrs)
>     518             assert isinstance(combined, Variable)
> 
> ValueError: 'userflag' is not present in all datasets.
> ```
> 
> If you can't reproduce this it may be that this only comes up when I re-run the cell (i.e. when it's grabbing from the object store?)

So this was due to a limited amount of metadata being used to lookup the Datasource for the data_level=1 and data_level=2 sources. This meant it was retrieving the same data for both. Now a set of required keys is used to lookup the Datasources. I've moved `datasource_lookup` into a function that can be called by each of the storage classes.

For an example see https://github.com/openghg/openghg/blob/5358aa86e467b4860d562b1e2bc5fab4906f6d36/openghg/store/_obssurface.py#L130